Step 2 - 요구사항 정리

min-queue opened the pull request:
안녕하세요 리뷰어님 설 잘보내셨나요? 
설에 지방에 있다 와서 조금 늦었습니다. 
리뷰 잘 부탁드립니다 리뷰어님 !! :)

min-queue committed:
docs: README.md 변경 및 메뉴 메뉴그룹 요구사항 정리

min-queue committed:
docs: README.md 상품 요구사항 정리

min-queue committed:
docs: README.md 주문 및 주문 테이블 요구사항 정리

min-queue committed:
docs: 클래스 다이어그램 추가

min-queue committed:
docs: README.md 에 클래스 다이어그램 추가

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+
+### 상품
+- 상품을 등록한다.
+  - [ ] 상품은 상품 그룹에 속해야 한다.
```
`상품 그룹`은 어떤 걸 의미할까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 상품은 상품 그룹에 속해야 한다.
+  - [ ] 상품 가격은 0원 이상이어야 한다.
+  - [ ] 상품 이름에는 비속어가 포함될 수 없다 (profanity).
+- 상품을 수정한다.
```
가격 외에 상품에 대해 수정하는 부분이 없으니
조금 더 구체적으로 작성해보면 좋을 것 같아요😊

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 상품 이름에는 비속어가 포함될 수 없다 (profanity).
+- 상품을 수정한다.
+  - [ ] 상품 가격은 있어야 하고, 0원 이상이어야 한다.
+  - [ ] 상품 가격 변경시 해당 상품을 포함한 메뉴의 가격도 변경된다.
```
상품 가격 변경 시 메뉴의 가격도 같이 변경될까요? 다시 확인해보면 좋을 것 같아요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+- 상품을 수정한다.
+  - [ ] 상품 가격은 있어야 하고, 0원 이상이어야 한다.
+  - [ ] 상품 가격 변경시 해당 상품을 포함한 메뉴의 가격도 변경된다.
+  - [ ] 상품 가격에 영향 받은 메뉴들 중 메뉴 가격이 새로운 가격보다 높은 경우 자동으로 노출을 해제한다.
```
`새로운 가격`이 변경할 상품의 가격이 맞을까요?
만약 맞다면 코드를 다시 확인해보면 좋을 것 같아요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 상품 가격 변경시 해당 상품을 포함한 메뉴의 가격도 변경된다.
+  - [ ] 상품 가격에 영향 받은 메뉴들 중 메뉴 가격이 새로운 가격보다 높은 경우 자동으로 노출을 해제한다.
+- 상품을 조회한다.
+  - [ ] 상품을 조회할 수 있다.
```
같은 내용의 경우 한 번만 정의하면 어떨까요? 🤔
주문, 주문 테이블 등 다른 조회 부분도 포함입니다!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+
+### 메뉴 그룹
+- 메뉴 그룹을 등록한다.
+  - [ ] 메뉴 그룹 이름은 필수이다.
```
메뉴 그룹 이름이 비어있는 경우도 등록할 수 있을까요?🤔

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 메뉴 가격은 0원 이상이어야 한다.
+  - [ ] 메뉴는 메뉴 그룹에 속해야 한다.
+  - [ ] 메뉴는 1개 이상의 상품으로 구성된다.
+  - [ ] 메뉴 생성 시 지정된 모든 상품이 실제로 데이터베이스에 존재해야 한다.
```
요구사항은 개발자뿐 아니라 다른 이해 관계자들도 충분히 이해할 수 있어야한다고 생각해요! `게시된 상품`, `등록된 상품`으로 풀어서 작성하면 어떨까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 메뉴는 1개 이상의 상품으로 구성된다.
+  - [ ] 메뉴 생성 시 지정된 모든 상품이 실제로 데이터베이스에 존재해야 한다.
+  - [ ] 메뉴 상품 수량은 0개 이상이어야 한다.
+  - [ ] 메뉴 가격의 합이 구성 상품의 가격 총합보다 큰 경우 등록할 수 없다.
```
17줄에는 메뉴 상품, 18줄에는 구성 상품이라고 표현하셨는데 표현을 통합하면 좋을 것 같아요! 그리고 `메뉴 가격의 합`이라고 표현하신 의도가 있을까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 메뉴 생성 시 지정된 모든 상품이 실제로 데이터베이스에 존재해야 한다.
+  - [ ] 메뉴 상품 수량은 0개 이상이어야 한다.
+  - [ ] 메뉴 가격의 합이 구성 상품의 가격 총합보다 큰 경우 등록할 수 없다.
+  - [ ] 메뉴 이름에는 비속어가 포함될 수 없다 (profanity).
```
(profanity)라고 추가로 표현하신 의도가 궁금해요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 메뉴 상품 수량은 0개 이상이어야 한다.
+  - [ ] 메뉴 가격의 합이 구성 상품의 가격 총합보다 큰 경우 등록할 수 없다.
+  - [ ] 메뉴 이름에는 비속어가 포함될 수 없다 (profanity).
+- 메뉴를 수정한다.
```
저는 메뉴에 대한 수정 사항들은 각각 독립적인 요구사항이라고 생각해요!
민규님은 수정으로 묶었던 의도가 궁금해요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - 메뉴 노출 상태 변경
+    [ ] 메뉴를 노출(display) 상태로 변경할 수 있다.
+      - 메뉴의 실제 가격이 구성 상품들의 총 가격보다 크면 노출 할 수 없다
+      - 메뉴의 실제 가격이 구성 상품들의 총 가격보다 작거나 같으면 노출 할 수 있다 
```
둘 중 하나만 표현해도 충분히 의미가 전달될 수 있을 것 같아요!
하나의 중복된 요소를 없애는 건 어떻게 생각하실까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+    [ ] 메뉴를 노출(display) 상태로 변경할 수 있다.
+      - 메뉴의 실제 가격이 구성 상품들의 총 가격보다 크면 노출 할 수 없다
+      - 메뉴의 실제 가격이 구성 상품들의 총 가격보다 작거나 같으면 노출 할 수 있다 
+    - [ ] 메뉴를 숨김(hide) 상태로 변경할 수 있다.
```
상품에서, `노출을 해제`라고 표현하셨는데 용어를 통일해보면 어떨까해요 😊
`노출 상태`, `숨김 상태`보다, `노출할 수 있다`, `숨길 수 있다`가 좀 더 다른 이해관계자가 보기도 편할 것 같다고도 생각되네요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  
+### 주문 테이블
+- 주문 테이블을 등록한다.
+  - [ ] 주문 테이블은 테이블 이름을 가지고 있어야 한다.
```
테이블 이름이 비어있는 경우도 등록할 수 있을까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+### 주문 테이블
+- 주문 테이블을 등록한다.
+  - [ ] 주문 테이블은 테이블 이름을 가지고 있어야 한다.
+  - [ ] 새로운 주문 테이블은 UUID 형식의 고유 식별자를 가지고 있어야 한다.
```
다른 이해관계자들이 UUID를 이해하기 어려울 것 같아요! 식별에 대한 내용이 요구사항에 필요할까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - 주문 테이블의 손님수 변경
+    - [ ] 주문 테이블이 존재해야 한다.
+    - [ ] 주문 테이블의 손님수는 0명 이상이어야 한다.
+    - [ ] 사용중인(occupied=true) 테이블의 손님 수만 변경할 수 있다/
```
`occupied=true`영역은 개발자가 아닌 다른 이해관계자가 볼 필요는 없다고 생각해요! 민규님은 어떻게 생각하실까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 상품을 조회할 수 있다.
+
+### 주문 
+- 주문을 생성한다.
```
누락된 정책 정책 확인 부탁드려요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 주문은 한 개 이상의 주문 상품을 가지고 있어야 한다.
+  - [ ] 주문 항목의 메뉴는 실제로 존재해야 한다.
+  - [ ] 매장 식사가 아닌 경우에는 수량이 음수일 수 없다.
+    - 매장식사의 경우는 음수수량을 허용(?)한다.
```
(?)는 빼셔도 될 것 같아요 ㅎㅎ

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - 주문을 서빙한다.
+    - [ ] 접수된 주문만 서빙할 수 있다.
+    - [ ] 주문 서빙 시 주문 상태는 서빙(SERVED)로 변경된다.
+  - 주문을 배달한다.
```
주문 상태와 상관없이 배달할 수 있을까요?

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - 주문을 배달한다.
+    - [ ] 배달 주문만 배달할 수 있다.
+    - [ ] 주문 배달 시 주문 상태는 배달(DELIVERING)로 변경된다.
+  - 주문을 완료한다.
```
배달 완료에 대한 요구사항이 누락된 것 같아요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+    - [ ] 배달 주문만 배달할 수 있다.
+    - [ ] 주문 배달 시 주문 상태는 배달(DELIVERING)로 변경된다.
+  - 주문을 완료한다.
+    - [ ] 주문 상태가 배달 완료(DELIVERED)된 주문만 완료할 수 있다.
```
`주문 타입이 배달인 경우`라는 전제가 붙으면 더 좋을 것 같아요

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+    - [ ] 주문 배달 시 주문 상태는 배달(DELIVERING)로 변경된다.
+  - 주문을 완료한다.
+    - [ ] 주문 상태가 배달 완료(DELIVERED)된 주문만 완료할 수 있다.
+    - [ ] 주문 타입이 매장식사(EAT_IN)와 테이크아웃(TAKEOUT)인 경우 주문 상태가 완료(COMPLETED)인 경우만 완료할 수 있다.
```
엇 상태 한 번 확인 부탁드릴게요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,98 @@
+  - [ ] 주문 테이블을 조회할 수 있다.
+
+### 모델링
+![kitchen pos](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/min-queue/ddd-legacy/step-2/class-diagram.puml)
```
모델링 잘해주셨습니다! ㅎㅎ
주문의 타입별로 상태에 대해서도 작성해보시면 어떨까요?!

kkt219a reviewed (changes requested):
안녕하세요 민규님! 설은 잘 보내셨나요 ㅎㅎ! 저도 지방을 갔다와서 조금 늦었네요ㅠㅠ 
2단계 미션 잘 진행해주셨어요! 🔥
몇가지 코멘트를 남겼는데 확인 부탁드려요!
+ 1단계 미션에서 마지막 코멘트 부분 반영도 확인 부탁드려요 ㅎㅎ

min-queue committed:
docs: updated uml file

min-queue committed:
docs: README.md 요구사항 업데이트

min-queue committed:
refact(stringCalculator):
- NumberValidator.java
  - Number에서의 상수 재활용
  - validateNumber메서드와 parseNumber를 동일한 행위로 변경
- Number.java ZERO를 외부에서 사용할수 없도록 변경

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+- 상품을 등록한다.
+  - [ ] 상품 가격은 0원 이상이어야 한다.
+  - [ ] 상품 이름에는 비속어가 포함될 수 없다 (profanity).
+- 상품을 가격을 수정한다.
+  - [ ] 상품 가격은 있어야 하고, 0원 이상이어야 한다.
+  - [ ] 메뉴 가격이 상품의 합산 가격보다 클때. 노출을 해제한다.
```
요구사항이 더 간결하고 확실해진 것 같아요! 👍🏻👍🏻

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+    - [ ] 주문 타입이 배달(DELIVERY)인 경우, 주문 상태가 배달 완료(DELIVERED)된 주문만 완료할 수 있다.
+    - [ ] 주문 타입이 매장식사(EAT_IN)와 테이크아웃(TAKEOUT)인 경우 주문 상태가 서빙 완료(SERVED)인 경우만 완료할 수 있다.
+    - [ ] 매장식사 주문이 완료되면 해당 테이블의 주문이 모두 완료된 경우 테이블을 초기화한다
+- [ ] 상품을 조회한다.
```
주문 테이블쪽도 그렇고 상품이라고 명시돼있어요🥲
메뉴쪽에 반영이 안된 부분도 있어서 전반적으로 확인 부탁드릴게요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+  - 주문을 서빙한다.
+    - [ ] 주문 상태가 접수(ACCEPTED)인 주문만 서빙할 수 있다.
+    - [ ] 주문 서빙 시 주문 상태는 서빙(SERVED)로 변경된다.
+  - 주문을 배달한다.
```
코드 영역을 제가 넓게 잡았었네요😅 요구사항 배달한다에 주문 상태 검증이 누락된 것 같아요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+  - 주문을 배달한다.
+    - [ ] 배달 주문만 배달할 수 있다.
+    - [ ] 주문 배달 시 주문 상태는 배달(DELIVERING)로 변경된다.
+  - 배달을 완료한다.
+    - [ ] 주문 상태가 배달 중(DELIVERING)인 주문만 배달 완료(DELIVERED)할 수 있다.
```
잘 반영해주셨습니다! 💯 💯

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+  - 배달을 완료한다.
+    - [ ] 주문 상태가 배달 중(DELIVERING)인 주문만 배달 완료(DELIVERED)할 수 있다.
+  - 주문을 완료한다.
+    - [ ] 주문 타입이 배달(DELIVERY)인 경우, 주문 상태가 배달 완료(DELIVERED)된 주문만 완료할 수 있다.
+    - [ ] 주문 타입이 매장식사(EAT_IN)와 테이크아웃(TAKEOUT)인 경우 주문 상태가 서빙 완료(SERVED)인 경우만 완료할 수 있다.
```
주문 타입에 따라 완료 가능한 조건이 다르죠! 잘 반영해주셨습니다 👍👍

kkt219a commented on README.md:
```diff
@@ -1,23 +1,107 @@
+#### 도메인 모델
+![kitchen pos](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/min-queue/ddd-legacy/step-2/uml/class-diagram.puml)
+
+#### 배달 주문 상태 다이어그램
+![배달 주문 상태 다이어그램](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/min-queue/ddd-legacy/step-2/uml/OrderType.DELIVERY.puml)
+
+#### 매장식사 주문 상태 다이어그램
+![매장식사 주문 상태 다이어그램](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/min-queue/ddd-legacy/step-2/uml/OrderType.EATIN.puml)
+
+#### 테이크아웃 주문 상태 다이어그램
+![테이크아웃 주문 상태 다이어그램](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/min-queue/ddd-legacy/step-2/uml/OrderType.TAKEOUT.puml)
```
주문 타입에 따른 상태가 따로 정리되어 있어 이해하기 훨씬 쉬운 것 같아요! 💯💯

kkt219a commented on src/main/java/calculator/Number.java:
```diff
@@ -2,17 +2,18 @@
 
 
 public class Number {
+    public static final Number ZERO = new Number(0);
```
불변 객체 보장 👍👍

kkt219a commented on src/main/java/calculator/NumberValidator.java:
```diff
@@ -1,20 +1,28 @@
+
     public void validate(String input) {
         if (!isBlank(input)) {
-            validateNumber(input);
-            validateNonNegative(parseNumber(input));
+            int parsedNumber = validateAndParseNumber(input);
+            validateNonNegative(parsedNumber);
         }
     }
```
검증과 파싱을 잘 묶어주셨어요 👍👍

kkt219a reviewed (changes requested):
안녕하세요 민규님 ! 😃
피드백 잘 반영해주셨어요! 👍🏻
진행하신 부분 코멘트 우선 남겼는데, DM 드린 부분이랑 함께 추가 확인 부탁드려요!😊

min-queue committed:
refact: 요구 사항 미반영 건 추가반영, 부족한 요구사항 추가

kkt219a commented on README.md:
```diff
@@ -1,23 +1,109 @@
+### 메뉴 그룹
+- 메뉴 그룹을 등록한다.
+  - [ ] 메뉴 그룹 이름은 빈 값이면 안되고, 필수이다.
+-  메뉴 그룹을 조회한다.
+  - [ ] 메뉴 그룹을 조회할 수 있다.
```
여기 중복된 부분만 확인 부탁드려요!

kkt219a commented on README.md:
```diff
@@ -1,23 +1,109 @@
+    - [ ] 주문 상태가 접수(ACCEPTED)인 주문만 서빙할 수 있다.
+    - [ ] 주문 서빙 시 주문 상태는 서빙(SERVED)로 변경된다.
+  - 주문을 배달한다.
+  - - [ ] 주문 상태가 서빙(SERVED)인 주문만 배달할 수 있다.
+    - [ ] 주문 타입이, 배달(DELIVERY)인 주문만 배달할 수 있다.
```
정책 잘 반영해주셨습니다 👍🏻👍🏻

kkt219a commented on README.md:
```diff
@@ -1,23 +1,109 @@
+- 메뉴를 등록한다.
+  - [ ] 메뉴 가격은 0원 이상이어야 한다.
+  - [ ] 메뉴는 메뉴 그룹에 속해야 한다.
+  - [ ] 메뉴는 1개 이상의 등록된 상품으로 구성된다.
+  - [ ] 메뉴 생성 시 지정된 모든 상품이 실제로 등록되어 있어야 한다.
```
다른 이해관계자도 잘 이해할 수 있을 것 같아요 👍🏻👍🏻

kkt219a reviewed (approved):
피드백 잘 반영해주셨어요 민규님! 💯
정정이 필요한 간단한 부분이 있는데, 다음 단계 진행하시면서 수정 부탁드릴게요 ㅎㅎ
이번 단계는 여기에서 머지하겠습니다! 다음 단계도 화이팅입니다 💪🏻💪🏻
